src/main/resources/liquibase/crop_changelog/23_1_0.xml

```diff
@@ -64,5 +64,337 @@
+        </preConditions>
+        <comment>
+            Call stored procedure fix_names_with_invalid_name_type for "fixing" names with invalid nametypes.
+        </comment>
+        <sql dbms="mysql" splitStatements="true">
+            CALL fix_names_with_invalid_name_type();
+        </sql>
+    </changeSet>
+
+    <changeSet author="mderamos" id="v23.1.0-5">
+        <preConditions onFail="MARK_RAN">
+            <sqlCheck expectedResult="1">
+                SELECT COUNT(*) FROM information_schema.routines
+                WHERE routine_schema = DATABASE() AND specific_name = 'fix_names_with_invalid_name_type';
+            </sqlCheck>
+        </preConditions>
+        <comment>Drop procedure fix_names_with_invalid_name_type</comment>
+        <sql dbms="mysql" splitStatements="false">
+            DROP PROCEDURE fix_names_with_invalid_name_type;
+        </sql>
+    </changeSet>
+
+    <changeSet author="mderamos" id="v23.1.0-6">
+        <preConditions onFail="MARK_RAN">
+            <and>
+                <sqlCheck expectedResult="0">
+                    SELECT COUNT(*) FROM information_schema.routines
+                    WHERE routine_schema = DATABASE() AND specific_name = 'enforce_unique_germplasm_name_type';
+                </sqlCheck>
+                <not>
+                    <sqlCheck expectedResult="0">
+                        SELECT COUNT(*) FROM
+                        (SELECT COUNT(*) FROM names GROUP BY gid, ntype HAVING COUNT(*) > 1) nametype_duplicates;
+                    </sqlCheck>
+                </not>
+            </and>
+        </preConditions>
+        <comment>
+            Create stored procedure enforce_unique_germplasm_name_type for enforcing unique germplasm's name nametypes
+            and creating new nametypes for the duplicates.
+        </comment>
+        <sql dbms="mysql" splitStatements="false">
+			<![CDATA[
+            CREATE PROCEDURE `enforce_unique_germplasm_name_type`()
+            BEGIN
+                DECLARE nameId INT;
+                DECLARE duplicateNameId INT;
+                DECLARE gid INT;
+                DECLARE nameTypeId INT;
+                DECLARE nameFtype VARCHAR(255);
+                DECLARE nameTable VARCHAR(255);
+                DECLARE nameTypeIndex INT DEFAULT 1;
+                DECLARE baseNameType VARCHAR(255);
+                DECLARE currentNameType VARCHAR(255);
+
+                DECLARE hasExistingNameType INT DEFAULT FALSE;
+                DECLARE isNameTypeUsedByGermplasm INT DEFAULT TRUE;
+                DECLARE done INT DEFAULT FALSE;
+
+                -- retrieved entries will not be updated
+                DECLARE names_with_duplicates CURSOR FOR
+                    SELECT name.nid as `nameId`, name.gid as `gid`, name.ntype as `nameTypeId`, nameType.ftype as `nameFtype`, nameType.ftable as `nameTable`
+                    FROM names name
+                    INNER JOIN udflds nameType ON nameType.fldno = name.ntype
```

Comment:
Also filter by ftype = 'NAME' to ensure only names are processed:

INNER JOIN udflds nameType ON nameType.fldno = name.ntype AND nameType.ftype = 'NAME'